src/app/clusters/bindings/BindingManager.h

```diff
@@ -0,0 +1,126 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/CASESessionManager.h>
+#include <app/server/Server.h>
+#include <app/util/binding-table.h>
+
+namespace chip {
+
+using BoundDeviceChangedHandler = void (*)(EndpointId localEndpoint, EndpointId remoteEndpoint, ClusterId clusterId,
+                                           OperationalDeviceProxy * peer_device);
+
+class BindingManager
+{
+public:
+    BindingManager() :
+        mOnConnectedCallback(HandleDeviceConnected, this), mOnConnectionFailureCallback(HandleDeviceConnectionFailure, this)
+    {}
+
+    void RegisterBoundDeviceChangedHandler(BoundDeviceChangedHandler handler) { mBoundDeviceChangedHandler = handler; }
+
+    void SetAppServer(Server * appServer) { mAppServer = appServer; }
+
+    CHIP_ERROR CreateBinding(FabricIndex fabric, NodeId node, EndpointId localEndpoint, ClusterId cluster);
+
+    CHIP_ERROR DisconnectDevice(FabricIndex fabric, NodeId node);
+
+    CHIP_ERROR NotifyBoundClusterChanged(EndpointId endpoint, ClusterId cluster);
+
+    static BindingManager & GetInstance() { return sBindingManager; }
+
+private:
+    static BindingManager sBindingManager;
+
+    static constexpr uint8_t kMaxPendingClusters = 5;
+
+    struct ClusterPath
+    {
+        EndpointId endpoint;
+        ClusterId cluster;
+    };
+
+    class PendingClusterEntry
+    {
+    public:
+        PeerId mPeerId;
```

Comment:
This split is odd - I would generally expect classes to not expose members and if we want members exposed to use structs instead, however  this class has one of the members private ... why? Can this be improved upon? Do we need class nesting here?